tools/bidder-bot/notifier/full.go

```diff
@@ -0,0 +1,146 @@
+package notifier
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+const (
+	BlockDuration = 12 * time.Second
+)
+
+type FullNotifier struct {
+	logger               *slog.Logger
+	targetBlockNumChan   chan uint64
+	l1Client             L1Client
+	notifySecondsAhead   time.Duration
+	lastNotifiedBlockNum uint64
+	mu                   sync.Mutex
+}
+
+func NewFullNotifier(
+	logger *slog.Logger,
+	l1Client L1Client,
+	notifySecondsAhead time.Duration,
+	targetBlockNumChan chan uint64,
+) *FullNotifier {
+	return &FullNotifier{
+		logger:             logger,
+		l1Client:           l1Client,
+		notifySecondsAhead: notifySecondsAhead,
+		targetBlockNumChan: targetBlockNumChan,
+	}
+}
+
+type L1Client interface {
+	SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error)
+}
+
+func (b *FullNotifier) Start(ctx context.Context) <-chan struct{} {
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+
+		headers := make(chan *types.Header)
+		sub, err := b.l1Client.SubscribeNewHead(ctx, headers)
+		if err != nil {
+			b.logger.Error("failed to subscribe to new heads", "error", err)
+			return
+		}
+
+		b.logger.Info("subscribed to new block headers")
+
+		for {
+			select {
+			case <-ctx.Done():
+				b.logger.Info("context done")
+				return
+
+			case err := <-sub.Err():
+				b.logger.Error("subscription error", "error", err)
+				sub, err = b.l1Client.SubscribeNewHead(ctx, headers)
+				if err != nil {
+					b.logger.Error("failed to resubscribe to new heads", "error", err)
+					return
+				}
+
+			case header := <-headers:
+				if err := b.HandleHeader(ctx, header); err != nil {
+					b.logger.Error("error handling header", "error", err)
+				}
+			}
+		}
+	}()
+	return done
+}
+
+func (b *FullNotifier) HandleHeader(ctx context.Context, header *types.Header) error {
```

Comment:
Why is this function exported?

Reply:
Just to test. Changed to export pattern in https://github.com/primev/mev-commit/commit/71ae460bfc9edc10f162526a04b370a39098f5db